src/transformers/models/phi/modular_phi.py

```diff
@@ -0,0 +1,1018 @@
+        self.num_attention_heads = num_attention_heads
+
+        if num_key_value_heads is None:
+            num_key_value_heads = num_attention_heads
+
+        self.num_key_value_heads = num_key_value_heads
+        self.resid_pdrop = resid_pdrop
+        self.embd_pdrop = embd_pdrop
+        self.attention_dropout = attention_dropout
+        self.hidden_act = hidden_act
+        self.max_position_embeddings = max_position_embeddings
+        self.initializer_range = initializer_range
+        self.layer_norm_eps = layer_norm_eps
+        self.use_cache = use_cache
+        self.rope_theta = rope_theta
+        self.rope_scaling = rope_scaling
+        self.partial_rotary_factor = partial_rotary_factor
+        self.qk_layernorm = qk_layernorm
+        # Validate the correctness of rotary position embeddings parameters
+        # BC: if there is a 'type' field, move it to 'rope_type'.
+        if self.rope_scaling is not None and "type" in self.rope_scaling:
+            self.rope_scaling["rope_type"] = self.rope_scaling["type"]
+        rope_config_validation(self)
+
+        super().__init__(
+            bos_token_id=bos_token_id,
+            eos_token_id=eos_token_id,
+            tie_word_embeddings=tie_word_embeddings,
+            **kwargs,
+        )
+
+
+class PhiRotaryEmbedding(LlamaRotaryEmbedding):
+    pass
+
+
+class PhiLinearScalingRotaryEmbedding(LlamaLinearScalingRotaryEmbedding, PhiRotaryEmbedding):
+    pass
+
+
+class PhiDynamicNTKScalingRotaryEmbedding(LlamaDynamicNTKScalingRotaryEmbedding, PhiRotaryEmbedding):
+    pass
+
+
+class PhiMLP(CLIPMLP):
+    pass
+
+
+class PhiAttention(GemmaAttention):
+    """Multi-headed attention from 'Attention Is All You Need' paper"""
+
+    def __init__(self, config: PhiConfig, layer_idx: Optional[int] = None):
+        super().__init__(config, layer_idx)
+        del self.o_proj
+        del self.scaling
+        del self.max_position_embeddings
+        self.head_dim = self.hidden_size // self.num_heads
+        self.rotary_ndims = int(self.head_dim * config.partial_rotary_factor)
+
+        self.q_proj = nn.Linear(self.hidden_size, self.num_heads * self.head_dim, bias=True)
+        self.k_proj = nn.Linear(self.hidden_size, self.num_key_value_heads * self.head_dim, bias=True)
+        self.v_proj = nn.Linear(self.hidden_size, self.num_key_value_heads * self.head_dim, bias=True)
+        self.dense = nn.Linear(self.num_heads * self.head_dim, self.hidden_size, bias=True)
+
+        self.qk_layernorm = config.qk_layernorm
+        if self.qk_layernorm:
+            self.q_layernorm = nn.LayerNorm(
+                config.hidden_size // self.num_heads, eps=config.layer_norm_eps, elementwise_affine=True
+            )
+            self.k_layernorm = nn.LayerNorm(
+                config.hidden_size // self.num_heads, eps=config.layer_norm_eps, elementwise_affine=True
+            )
+
+        self.rotary_emb = PhiRotaryEmbedding(config=self.config)
```

Comment:
There is no reason to inherit from `GemmaAttention` here IMO, it just makes things harder to understand (we redefine everything anyway). Since there are no close match to `PhiAttention`, let's just inherit from `nn.Module` and use the original code.

Reply:
Oh Yes. Agreed. Makes it more complicated, (Modularity is for simplifying). Initially kept it that way.

Although it seems to save few lines of code but at the cost of Simplicity.
Rolling back to base classing with `nn.Module`